File: project-{topic}-optimized/src/main/java/com/alx/devops/controller/ProductController.java
package com.alx.devops.controller;

import com.alx.devops.dto.ProductDTO;
import com.alx.devops.service.ProductService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST Controller for managing products.
 * Provides CRUD operations for products with role-based access control and search functionality.
 */
@RestController
@RequestMapping("/api/products")
@RequiredArgsConstructor
@Tag(name = "Product Management", description = "APIs for managing products")
public class ProductController {

    private final ProductService productService;

    /**
     * Retrieves all products or searches for products by a keyword. Accessible by all authenticated users.
     *
     * @param search An optional keyword to search for in product names or descriptions.
     * @return A list of ProductDTOs.
     */
    @Operation(summary = "Get all products or search by keyword")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successfully retrieved list of products"),
            @ApiResponse(responseCode = "401", description = "Unauthorized - Authentication required")
    })
    @GetMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
    public ResponseEntity<List<ProductDTO>> getAllProducts(
            @Parameter(description = "Optional search keyword for product name or description") @RequestParam(required = false) String search) {
        if (search != null) {
            return ResponseEntity.ok(productService.searchProducts(search));
        }
        List<ProductDTO> products = productService.getAllProducts();
        return ResponseEntity.ok(products);
    }

    /**
     * Retrieves a single product by its ID. Accessible by all authenticated users.
     *
     * @param id The ID of the product.
     * @return The ProductDTO if found.
     */
    @Operation(summary = "Get a product by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Product found"),
            @ApiResponse(responseCode = "404", description = "Product not found"),
            @ApiResponse(responseCode = "401", description = "Unauthorized - Authentication required")
    })
    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
    public ResponseEntity<ProductDTO> getProductById(
            @Parameter(description = "ID of the product to retrieve") @PathVariable Long id) {
        ProductDTO product = productService.getProductById(id);
        return ResponseEntity.ok(product);
    }

    /**
     * Creates a new product. Only accessible by ADMIN users.
     *
     * @param productDTO The ProductDTO containing new product data.
     * @return The created ProductDTO.
     */
    @Operation(summary = "Create a new product (Admin only)")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Product created successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid input or category not found"),
            @ApiResponse(responseCode = "401", description = "Unauthorized - Authentication required"),
            @ApiResponse(responseCode = "403", description = "Forbidden - Requires ADMIN role")
    })
    @PostMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ProductDTO> createProduct(
            @Parameter(description = "Product object to create") @Valid @RequestBody ProductDTO productDTO) {
        ProductDTO createdProduct = productService.createProduct(productDTO);
        return new ResponseEntity<>(createdProduct, HttpStatus.CREATED);
    }

    /**
     * Updates an existing product. Only accessible by ADMIN users.
     *
     * @param id         The ID of the product to update.
     * @param productDTO The ProductDTO with updated data.
     * @return The updated ProductDTO.
     */
    @Operation(summary = "Update an existing product (Admin only)")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Product updated successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid input or category not found"),
            @ApiResponse(responseCode = "404", description = "Product not found"),
            @ApiResponse(responseCode = "401", description = "Unauthorized - Authentication required"),
            @ApiResponse(responseCode = "403", description = "Forbidden - Requires ADMIN role")
    })
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ProductDTO> updateProduct(
            @Parameter(description = "ID of the product to update") @PathVariable Long id,
            @Parameter(description = "Updated product object") @Valid @RequestBody ProductDTO productDTO) {
        ProductDTO updatedProduct = productService.updateProduct(id, productDTO);
        return ResponseEntity.ok(updatedProduct);
    }

    /**
     * Deletes a product by its ID. Only accessible by ADMIN users.
     *
     * @param id The ID of the product to delete.
     * @return A ResponseEntity with no content.
     */
    @Operation(summary = "Delete a product (Admin only)")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204", description = "Product deleted successfully"),
            @ApiResponse(responseCode = "404", description = "Product not found"),
            @ApiResponse(responseCode = "401", description = "Unauthorized - Authentication required"),
            @ApiResponse(responseCode = "403", description = "Forbidden - Requires ADMIN role")
    })
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteProduct(
            @Parameter(description = "ID of the product to delete") @PathVariable Long id) {
        productService.deleteProduct(id);
        return ResponseEntity.noContent().build();
    }
}

File: project-{topic}-advanced/src/controllers/product.controller.js
```javascript
const httpStatus = require('http-status');
const pick = require('../utils/pick');
const ApiError = require('../utils/ApiError');
const catchAsync = require('../utils/catchAsync');
const productService = require('../services/product.service');
const logger = require('../utils/logger');

/**
 * Creates a new product. Only accessible by admins.
 */
const createProduct = catchAsync(async (req, res) => {
  const product = await productService.createProduct(req.body);
  logger.info(`Admin (ID: ${req.user.id}) created new product: ${product.name} (ID: ${product.id})`);
  res.status(httpStatus.CREATED).send(product);
});

/**
 * Retrieves multiple products with pagination and filtering. Accessible by all authenticated users.
 */
const getProducts = catchAsync(async (req, res) => {
  const filter = pick(req.query, ['name']);
  const options = pick(req.query, ['sortBy', 'limit', 'page']);
  const result = await productService.queryProducts(filter, options);
  logger.debug(`User (ID: ${req.user.id}) fetched products with filter: ${JSON.stringify(filter)}`);
  res.send(result);
});

/**
 * Retrieves a single product by ID. Accessible by all authenticated users.
 */
const getProduct = catchAsync(async (req, res) => {
  const product = await productService.getProductById(req.params.productId);
  if (!product) {
    throw new ApiError(httpStatus.NOT_FOUND, 'Product not found');
  }
  logger.debug(`User (ID: ${req.user.id}) fetched product with ID: ${product.id}`);
  res.send(product);
});

/**
 * Updates a product by ID. Only accessible by admins.
 */
const updateProduct = catchAsync(async (req, res) => {
  const product = await productService.updateProductById(req.params.productId, req.body);
  logger.info(`Admin (ID: ${req.user.id}) updated product with ID: ${product.id}`);
  res.send(product);
});

/**
 * Deletes a product by ID. Only accessible by admins.
 */
const deleteProduct = catchAsync(async (req, res) => {
  await productService.deleteProductById(req.params.productId);
  logger.info(`Admin (ID: ${req.user.id}) deleted product with ID: ${req.params.productId}`);
  res.status(httpStatus.NO_CONTENT).send();
});

module.exports = {
  createProduct,
  getProducts,
  getProduct,
  updateProduct,
  deleteProduct,
};
```